### PR TITLE
Implements DreamList unions (aka `L1 | L2`)

### PR DESCRIPTION
**Pull request by ZeWaka**
title

partially solves #120 - won't be doing the icon bit now

```cs
var/list/L = list(1,2,3)
var/list/G = list(3,4,5)
var/x = 7
var/y = 8
world.log << json_encode(L | G)
world.log << json_encode(G | L)
world.log << json_encode(L | x)
world.log << json_encode(x | y)
world.log << json_encode(L | null)
world.log << json_encode(null | L)
```
to
```
[1,2,3,4,5]
[3,4,5,1,2]
[1,2,3,7]
15
[1,2,3,null]
null
```

### CODE REVIEW

File: OpenDreamRuntime/Objects/DreamList.cs [hunk 1]
```diff
@@ -144,6 +144,15 @@ public class DreamList : DreamObject {
         public int GetLength() {
             return _values.Count;
         }
+
+        public DreamList Union(DreamList other) {
```
Comment (ZeWaka) [inline OpenDreamRuntime/Objects/DreamList.cs:148]: could also override the `|` operator if you want

File: OpenDreamRuntime/Procs/DMOpcodeHandlers.cs [hunk 2]
```diff
@@ -526,10 +526,16 @@ static class DMOpcodeHandlers {
...
         public static ProcStatus? BitOr(DMProcState state) {
-            int second = state.PopDreamValue().GetValueAsInteger();
-            int first = state.PopDreamValue().GetValueAsInteger();
+            DreamValue value = state.PopDreamValue();
+            if (value.TryGetValueAsInteger(out int secondInt)) {
+                int first = state.PopDreamValue().GetValueAsInteger();
+
+                state.Push(new DreamValue(first | secondInt));
+            } else if (value.TryGetValueAsDreamList(out DreamList secondList)) {
+                DreamList first = state.PopDreamValue().GetValueAsDreamList();
```
Comment (ike709) [inline OpenDreamRuntime/Procs/DMOpcodeHandlers.cs:535]: What if someone mixes an integer and a list?

Comment (ZeWaka) [reply OpenDreamRuntime/Procs/DMOpcodeHandlers.cs:535]: :sadcat:

File: OpenDreamRuntime/Procs/DMOpcodeHandlers.cs [hunk 3]
```diff
@@ -525,11 +525,35 @@ static class DMOpcodeHandlers {
...
+        public static ProcStatus? BitOr(DMProcState state) {                                // x | y
+            DreamValue secondValue = state.PopDreamValue();
 
-            state.Push(new DreamValue(first | second));
+            if (secondValue.TryGetValueAsInteger(out int secondInt)) {                      // x | 2
+                DreamValue firstValue = state.PopDreamValue();
+
+                if (firstValue.TryGetValueAsInteger(out int firstInt)) {                    // 1 | 2
+                    state.Push(new DreamValue(firstInt | secondInt));
+                } else if (firstValue.TryGetValueAsDreamList(out DreamList firstList)) {    // List | 2
```
Comment (wixoaGit) [inline OpenDreamRuntime/Procs/DMOpcodeHandlers.cs:536]: How this is done with other operators (`+`, `-`, etc) is there's an operator method in `IDreamMetaObject`. If the first value is a `DreamObject`, that method gets called. The override of `OperatorOr` in `DreamMetaObjectList` would handle the union.

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**